### PR TITLE
AB#60-Put-transaction-to-neo4j-database

### PR DESCRIPTION
**Pull request by steboee**
#60 - Put-transaction-to-neo4j-database
- updated code structure
- added console menu
- pushing transaction to neo4j database
- added basic examples of neomodels
- removed unnecessary files

### CODE REVIEW

File: src/populate_neo.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    total_items = len(data)
+    processed_items = 0
+    for item in data:
+        transaction = Transaction(
+            from_address=item['fromAddress'],
+            to_address=item['toAddress'],
+            transaction_id=item['id'],
+            nft=item['nft']['id']
+        )
+        transaction_service.processTransaction(transaction)
```
Comment (brano-hozza) [inline src/populate_neo.py:35]: This should be called only once

File: .gitignore [hunk 2]
```diff
@@ -6,4 +6,8 @@ __pycache__/
 .env
 venv/
 
-*.log
\ No newline at end of file
+*.log
+
+# ignore the vscode
+.vscode/
```
Comment (brano-hozza) [inline .gitignore:12]: We can keep this workspace config in repo

File: .gitignore [hunk 3]
```diff
@@ -6,4 +6,8 @@ __pycache__/
...
+```
```
Comment (brano-hozza) [inline .gitignore:13]: Remove

File: main.py [hunk 4]
```diff
@@ -1,56 +1,20 @@
...
-    headers={
-        "Content-Type": "application/json",
-        "Accept": "application/json"
-    }
-)
+    neo4j = Neo4jDatabase(neo4j_uri, neo4j_user, neo4j_password)
+    logfile = DataProcessingLogger("src/logs/indexer_data/indexer.log")
 
-parsed = raw.json()
+    with neo4j.driver.session() as session:
```
Comment (brano-hozza) [inline main.py:17]: Do we need this session scope?

File: src/fetch_data.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+
+
+
+def main():
+    logger = DataProcessingLogger("src/logs/indexer_data/indexer.log")
+
+    indexer_uri = os.getenv("INDEXER_URI")
+    raw = requests.post(indexer_uri+"/graphql", json = {
+        "operationName": "getTransactions",
```
Comment (brano-hozza) [inline src/fetch_data.py:13]: We could move this constant graphql strings to some different file

File: src/models/neo4j_models.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+from neomodel import StructuredNode, StringProperty, RelationshipTo, Relationship
+import os
+from neomodel import config
+from dotenv import load_dotenv
+
+
+load_dotenv()
```
Comment (brano-hozza) [inline src/models/neo4j_models.py:7]: This should be called only once at the start of the program

File: src/populate_neo.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+#SCRIPT TO INSERT FIRST DATA FROM LOG FILE TO NEO4J
+def main():
+    indexer_log = DataProcessingLogger.get_instance().log_file
+    data = []
+    with open(indexer_log, 'r') as file:
+        for line in file:
+            data.append(json.loads(line))
+
+    load_dotenv()
```
Comment (brano-hozza) [inline src/populate_neo.py:18]: Again

File: src/services/TransactionService.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        with self.driver.session() as session:
+            session.write_transaction(self._create_sent_relationship, from_address, to_address)
+            session.write_transaction(self._create_received_relationship, from_address, to_address)
+            session.write_transaction(self._insert_nft, transaction, nft)
+            session.write_transaction(self._create_nft_relationship, transaction, nft)
+
+
+    # UTIL FUNCTION, IS ADDRESS IN GRAPH ?
+    @staticmethod  
+    def _is_address_in_graph(tx, address):
```
Comment (brano-hozza) [inline src/services/TransactionService.py:41]: I would move this static methods to something like TransactionRepository class, so we keep only domain logic in service

**Review by brano-hozza — CHANGES_REQUESTED** [review]

File: src/services/TransactionService.py (hunk 8 above)
Comment (steboee) [reply src/services/TransactionService.py]: addded TransactionRepository where all static methods are stored now

File: src/populate_neo.py (hunk 1 above)
Comment (steboee) [reply src/populate_neo.py:32]: yeah my bad, (tired at night doing this)

File: .gitignore (hunk 2 above)
Comment (steboee) [reply .gitignore]: okay,

File: main.py (hunk 4 above)
Comment (steboee) [reply main.py]: nope, it remained there while refactoring

File: src/fetch_data.py (hunk 5 above)
Comment (steboee) [reply src/fetch_data.py:12]: i created basic IndexerService where all logic will be (in future)

File: src/fetch_data.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                            id
+                            }
+                        }
+                    }"""
+        }
+
+    headers = {
+         "Content-Type": "application/json",
+            "Accept": "application/json"
+    }
```
Comment (brano-hozza) [inline src/fetch_data.py:30]: I don't think headers would change so we can put those also in the runQuery method withou passing them

**Review by brano-hozza — CHANGES_REQUESTED** [review]

**Review by brano-hozza — APPROVED** [review]

**Pull request merged**